### PR TITLE
feat(abci): add a custom process proposal 

### PR DESCRIPTION
**Pull request by itsdevbear**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new configuration option "IsMainnet" for the "Polar" settings.
	- Added a new "ProcessProposal" function to process proposals and prepare the chain state.
	- Introduced a new "WrappedBlockchain" struct with methods for writing the genesis state and inserting blocks.
- **Bug Fixes**
	- Modified the "EndBlock" function to obtain the block variable from a different chain implementation.
	- Updated the "InitGenesis" function to use the new "WrappedBlockchain" interface.
- **Refactor**
	- Replaced the "Blockchain" interface with the new "WrappedBlockchain" interface in the "Keeper" struct.
	- Updated import paths and test descriptions in various files.
- **Documentation**
	- Added comments to highlight issues with state changes in the "PreBlock" and "BeginBlock" functions.

### CODE REVIEW

**Review by ocnc — APPROVED** [review]

**Review by garrettparris — APPROVED** [review]

**Review by BrickBera — APPROVED** [review]

File: eth/core/chain_writer.go [hunk 1]
```diff
@@ -50,36 +51,56 @@ func (bc *blockchain) WriteGenesisBlock(block *types.Block) error {
...
+}
+
+// insertBlockWithoutSetHead inserts a block into the blockchain without setting it as the head.
+func (bc *blockchain) insertBlockWithoutSetHead(
+	block *types.Block,
+) ([]*types.Receipt, []*types.Log, error) {
 	// Validate that we are about to insert a valid block.
+	// If the block number is greater than 1,
+	// it means it's not the genesis block and needs to be validated. TODO kinda hood.
 	if block.NumberU64() > 1 { // TODO DIAGNOSE
```
Comment (nivasan1) [inline eth/core/chain_writer.go:69]: What is the purpose of this conditional?

**Review by nivasan1 — COMMENTED** [review]
Comments + questions. TLDR is that ensure that the execution env of Process / Prepare are the same. Otherwise, you may introduce liveness failures

File: eth/core/chain_writer.go (hunk 1 above)
Comment (itsdevbear) [reply eth/core/chain_writer.go:69]: a bug we can't figure out yet LOL

File: cosmos/runtime/chain/abci.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	ctx.BlockGasMeter().RefundGas(ctx.BlockGasMeter().GasConsumed(), "prepare proposal")
+	ctx = ctx.WithKVGasConfig(storetypes.GasConfig{}).
+		WithTransientKVGasConfig(storetypes.GasConfig{}).
+		WithGasMeter(storetypes.NewInfiniteGasMeter())
+
+	// Pull an execution payload out of the proposal.
+	var envelope *engine.ExecutionPayloadEnvelope
+	for _, tx := range req.Txs {
+		var sdkTx sdk.Tx
+		sdkTx, err = wbc.app.TxDecode(tx)
```
[the remark this answers is not in the document]
Comment (itsdevbear) [reply cosmos/runtime/chain/abci.go:74]: addressed

File: cosmos/runtime/chain/abci.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+func (wbc *WrappedBlockchain) ProcessProposal(
+	ctx sdk.Context, req *abci.RequestProcessProposal,
+) (*abci.ResponseProcessProposal, error) {
+	var (
+		err error
+	)
+
+	// We have to run the PreBlocker && BeginBlocker to get the chain into the state
+	// it'll be in when the EVM transaction actually runs.
+	if _, err = wbc.app.PreBlocker(ctx, &abci.RequestFinalizeBlock{
```
[the remark this answers is not in the document]
Comment (itsdevbear) [reply cosmos/runtime/chain/abci.go:47]: they will be identical here yup

**Pull request merged**